---
title: "ACORN Pool-Seq Allele frequencies validation"
author: "Nikolaos Tourvas & Lars Littmann"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    code_download: yes
    code_folding: show
    theme: default
    toc: yes
    toc_float: yes
    keep_md: no
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```

Load libraries
```{r}
library(Metrics)
library(tidyverse)
library(patchwork)
library(viridis)
library(parallel)
```

Functions
```{r}
getpooldata <- function(tablepath, pool1, pool2){
  pooldata <- read.table(tablepath, 
                         sep = "\t", header=TRUE, check.names = FALSE)
  pool1 <- pooldata %>% 
                     select(chrom_pos, `pool1`)
  colnames(pool1) <- c("site", "alt_pool")
  pool1 <- pool1 %>% 
                     filter(alt_pool > 0)
  
  pool2 <- pooldata %>% 
                     select(chrom_pos, `pool2`)
  colnames(pool2) <- c("site", "alt_pool")
  pool2 <- pool2 %>% 
                     filter(alt_pool > 0)
  
  # Make consensus data set
  consensuspool1 <- inner_join(inds305, pool1)
  consensuspool2 <- inner_join(inds312, pool2)
  
  # Add minor allele frequency (MAF) column
  consensuspool1$maf_ind <- ifelse(
    consensuspool1$alt_ind <= 0.5,
    consensuspool1$alt_ind, 1 - consensuspool1$alt_ind)

  consensuspool2$maf_ind <- ifelse(
    consensuspool2$alt_ind <= 0.5,
    consensuspool2$alt_ind, 1 - consensuspool2$alt_ind)

  consensuspool1$maf_pool <- ifelse(
    consensuspool1$alt_pool <= 0.5,
    consensuspool1$alt_pool, 1 - consensuspool1$alt_pool)

  consensuspool2$maf_pool <- ifelse(
    consensuspool2$alt_pool <= 0.5,
    consensuspool2$alt_pool, 1 - consensuspool2$alt_pool)

  poplist <- list(consensuspool1, consensuspool2)
  
  return(poplist)
}

error_stats <- function(ind_freq, pool_freq){
    # Fit a linear regression model
    model <- lm(ind_freq ~ pool_freq)
    
    # Pearson's r
    r <- paste0("r = ", cor(ind_freq, pool_freq, method = "pearson"))
    
    # Extract R-squared value from the model summary
    #Rsquared<- paste0("R squared = ",summary(model)$r.squared)
    
    # rmse
    #rmse <- paste0("RMSE = ",rmse(ind_freq, pool_freq))
    
    # mae
    mae <- paste0("MAE = ",mae(ind_freq, pool_freq))
    
    # mape (Mean Absolute Percentage Error)
  #   mape <- function(y_true, y_pred) {
  #     non_zero_indices <- which(y_true != 0)
  #     if (length(non_zero_indices) > 0) {
  #       mean(abs((y_true[non_zero_indices] - y_pred[non_zero_indices]) / y_true[non_zero_indices])) * 100
  #     } else {
  #       NA
  #     }
  #   }
  #   mape_res <- paste0("MAPE = ",mape(ind_freq, pool_freq))
  out <- paste("\n", r, mae, sep="\n")
  return(out)
}

hist2d <- function(dataset, alt_ind, alt_pool, bins=39){
#Create the 2D histogram plot:
  p <- ggplot(dataset, aes(x=alt_ind, y=alt_pool)) + 
       geom_bin2d(bins=bins) + 
       scale_fill_viridis(trans='log10') +
       theme_minimal() + 
       labs(x="Ind-Seq ALT frequency",
            y="Pool-Seq ALT frequency",
            fill="Log10 Density") + 
       theme(legend.position = c(1.2, 1.15))
  
  #Create marginal histograms:
  px <- ggplot(dataset, aes(x=alt_ind)) + 
    geom_density(color="black", fill="lightgrey", adjust=0.5) + theme_void()
  py <- ggplot(dataset, aes(x=alt_pool)) +
    geom_density(color="black", fill="lightgrey", adjust=0.5) + theme_void() +
                coord_flip()
  
  #Arrange the plots together:
  px + plot_spacer() + p + py +
    plot_layout(
      ncol = 2, nrow = 2,
      widths = c(4,1), heights = c(1,4)
    ) +
    plot_annotation(title = title)
}

# hist2dmaf <- function(dataset, maf_ind, maf_pool){
# #Create the 2D histogram plot:
#   p <- ggplot(dataset, aes(x=maf_ind, y=maf_pool)) + 
#        geom_bin2d(bins=20) + 
#        scale_fill_viridis(trans='log10') +
#        theme_minimal() + 
#        labs(x="Ind-Seq Minor Allele Frequency",
#             y="Pool-Seq Minor Allele Frequency",
#             fill="Log10 Density") + 
#        theme(legend.position = c(1.2, 1.15))
#   
#   #Create marginal histograms:
#   px <- ggplot(dataset, aes(x=maf_ind)) + geom_histogram(bins=21) + theme_void()
#   py <- ggplot(dataset, aes(x=maf_pool)) + geom_histogram(bins=21) + theme_void() +
#                 coord_flip()
#   
#   #Arrange the plots together:
#   px + plot_spacer() + p + py +
#     plot_layout(
#       ncol = 2, nrow = 2,
#       widths = c(4,1), heights = c(1,4)
#     ) +
#     plot_annotation(title = title)
# }

rawdiffplot <- function(dataset, bins=37){
  p <- ggplot(dataset, aes(x=alt_pool - alt_ind, y=maf_pool)) +
       geom_bin2d(bins=100) + 
       scale_fill_viridis(trans='log10') +
       theme_minimal() + 
       labs(x="Raw AF differences (Pool AF - Ind AF)",
            y="Pool-Seq Minor Allele Frequency",
            fill="Log10 Density") + 
       theme(legend.position = c(1.2, 1.15))
  #Create marginal histograms:
  px <- ggplot(dataset, aes(x=alt_pool - alt_ind)) + 
          geom_density(color="black", fill="lightgrey", adjust=0.5) + theme_void()
  py <- ggplot(dataset, aes(x=alt_ind)) + 
          geom_density(color="black", fill="lightgrey", adjust=0.5) + theme_void() + coord_flip()
  
  #Arrange the plots together:
  px + plot_spacer() + p + py +
    plot_layout(
      ncol = 2, nrow = 2,
      widths = c(4,1), heights = c(1,4)
    ) +
    plot_annotation(title = title)
}
  
histALTfreq <- function(dataset, alt_ind, alt_pool, title="", bins=41){
  ggplot(dataset) +
  geom_histogram(aes(x=alt_ind, after_stat(density)), bins = bins,
                 fill = "lightblue", color = "black") +
  geom_density(aes(x=alt_pool), color="darkred", linewidth=1) +
  geom_vline(xintercept = .05, color="black", linetype="dashed") +
  theme_minimal() +  
  labs(x="ALT frequency", y="Density",
       title=title)
}

# histMAFfreq <- function(dataset, maf_ind, maf_pool, title=""){
#   ggplot(dataset) +
#   geom_histogram(aes(x=maf_ind, after_stat(density)), bins = 21,
#                  fill = "lightblue", color = "black") +
#   geom_density(aes(x=maf_pool), color="darkred", linewidth=1) +
#   geom_vline(xintercept = .05, color="black", linetype="dashed") +
#   theme_minimal() +  
#   labs(x="Minor Allele Frequency", y="Density",
#        title=title)
# }

AF_barplot <- function(dataset, title="", bins=41){
  dataset <- dataset[ ,-c(4,5)]
  colnames(dataset) <- c("site", "Ind-Seq", "Pool-Seq")
  dataset_tidy <- pivot_longer(dataset, names_to="seq_type",
                                 values_to = "freq", !site)
  
  ggplot(dataset_tidy, aes(x=freq, fill=seq_type))+
    geom_histogram(position = "dodge", bins = bins, color="black") +
    geom_vline(xintercept = .05, color="black", linetype="dashed") +
    scale_fill_manual(values = c("lightblue", "darkred")) +
    theme_minimal() +
    theme(legend.position = c(.8, .9)) +
    labs(x = "ALT Frequency", fill = "Sequencing method", title=title)
}
```

Load individuals data
```{r}
# Inds data
inds305 <- read.table("/data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/AF_tables/ACORN_VCF_Individual_Pool305_AlleleFreqs.frq", header=TRUE, 
                   row.names = NULL, na.strings = "-nan")

colnames(inds305) <- c("chrom", "pos", "alleles", "read_depth", "ref_ind", "alt_ind")
inds305 <- inds305 %>% 
  unite(site, chrom, pos, sep="_")

## Filter for coverage and allele frequency AF>0
## The second filter is applied because we did the variant calling for
## both pops together. That's the only reason why we would get AF=0
inds305 <- inds305 %>% 
  filter(read_depth >= 40) %>% 
  filter(alt_ind > 0) %>% 
  select(site, alt_ind)

inds312 <- read.table("/data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/AF_tables/ACORN_VCF_Individual_Pool312_AlleleFreqs.frq", sep="\t", header=TRUE, 
                   row.names = NULL, na.strings = "-nan")

colnames(inds312) <- c("chrom", "pos", "alleles", "read_depth", "ref_ind", "alt_ind")
inds312 <- inds312 %>% 
  unite(site, chrom, pos, sep="_")

## Filter for coverage and allele frequency AF>0
## The second filter is applied because we did the variant calling for
## both pops together. That's the only reason why we would get AF=0
inds312 <- inds312 %>% 
  filter(read_depth >= 40) %>% 
  filter(alt_ind > 0) %>% 
  select(site, alt_ind)
```

# Standard samtools markdup
Load pool data
```{r}
# Pool data
poplist <- getpooldata("/data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100_AlleleFrequencyTable.txt", "305", "312")
poplist_markdup <- poplist
```

Data set size
```{bash, eval=FALSE}
bcftools stats --threads 10 --fasta-ref /data/genetics_tmp/TEST_LL/REFERENCE/Qrob_PM1N_Organelles.fa \
    /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf.gz \
    > /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Pool VCF (populations 305 and 312):
```{bash}
grep "number of SNPs:" /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Consensus data sets (per population inner join of Ind VCF and Pool VCF):
```{r}
paste0("Number of SNPs in consensus data set of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_markdup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Standard duplicate removal"
# print the results
cat(paste0(title,results_markdup[[1]], results_markdup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   hist2dmaf(dataset=x)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histMAFfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

Number of SNPs in the Consensus data sets after MAF>0.05 (per population inner join of Ind VCF and Pool VCF):
```{r}
poplist <- poplist
for (i in 1:2){
  poplist[[i]][,c(4,5)][poplist[[i]][,c(4,5)] < 0.05] <- NA
  poplist[[i]] <- na.omit(poplist[[i]])
}

paste0("Number of SNPs in consensus data set (MAF>0.05) of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set (MAF>0.05) of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_markdup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Standard duplicate removal - MAF>0.05"
# print the results
cat(paste0(title,results_markdup[[1]], results_markdup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x, bins=36)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title, bins = 39)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```

# Deduplication with fastp
Load pool data
```{r}
# Pool data
poplist <- getpooldata("/data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100_AlleleFrequencyTable.txt", "305_dedup", "312_dedup")
poplist_dedup <- poplist
```

Data set size
```{bash, eval=FALSE}
bcftools stats --threads 10 --fasta-ref /data/genetics_tmp/TEST_LL/REFERENCE/Qrob_PM1N_Organelles.fa \
    /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf.gz \
    > /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Pool VCF (populations 305 and 312):
```{bash}
grep "number of SNPs:" /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Consensus data sets (per population inner join of Ind VCF and Pool VCF):
```{r}
paste0("Number of SNPs in consensus data set of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_dedup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Deduplication with fastp"
# print the results
cat(paste0(title,results_dedup[[1]], results_dedup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   hist2dmaf(dataset=x)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histMAFfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

Number of SNPs in the Consensus data sets after MAF>0.05 (per population inner join of Ind VCF and Pool VCF):
```{r}
poplist <- poplist
for (i in 1:2){
  poplist[[i]][,c(4,5)][poplist[[i]][,c(4,5)] < 0.05] <- NA
  poplist[[i]] <- na.omit(poplist[[i]])
}

paste0("Number of SNPs in consensus data set (MAF>0.05) of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set (MAF>0.05) of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_markdup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Deduplication with fastp - MAF>0.05"
# print the results
cat(paste0(title,results_markdup[[1]], results_markdup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x, bins=36)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title, bins = 39)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```

# Standard samtools markdup RD>=30
Load pool data
```{r}
# Pool data
poplist <- getpooldata("/data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100_AlleleFrequencyTable.txt", "305", "312")
poplist_markdup <- poplist
```

Data set size
```{bash, eval=FALSE}
bcftools stats --threads 10 --fasta-ref /data/genetics_tmp/TEST_LL/REFERENCE/Qrob_PM1N_Organelles.fa \
    /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf.gz \
    > /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Pool VCF (populations 305 and 312):
```{bash}
grep "number of SNPs:" /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_markdup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Consensus data sets (per population inner join of Ind VCF and Pool VCF):
```{r}
paste0("Number of SNPs in consensus data set of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_markdup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Standard duplicate removal RD>=30"
# print the results
cat(paste0(title,results_markdup[[1]], results_markdup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   hist2dmaf(dataset=x)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histMAFfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

Number of SNPs in the Consensus data sets after MAF>0.05 (per population inner join of Ind VCF and Pool VCF):
```{r}
poplist <- poplist
for (i in 1:2){
  poplist[[i]][,c(4,5)][poplist[[i]][,c(4,5)] < 0.05] <- NA
  poplist[[i]] <- na.omit(poplist[[i]])
}

paste0("Number of SNPs in consensus data set (MAF>0.05) of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set (MAF>0.05) of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_markdup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Standard duplicate removal RD>=30 - MAF>0.05"
# print the results
cat(paste0(title,results_markdup[[1]], results_markdup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x, bins=36)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title, bins = 39)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```

# Deduplication with fastp RD>=30
Load pool data
```{r}
# Pool data
poplist <- getpooldata("/data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100_AlleleFrequencyTable.txt", "305_dedup", "312_dedup")
poplist_dedup <- poplist
```

Data set size
```{bash, eval=FALSE}
bcftools stats --threads 10 --fasta-ref /data/genetics_tmp/TEST_LL/REFERENCE/Qrob_PM1N_Organelles.fa \
    /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf.gz \
    > /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Pool VCF (populations 305 and 312):
```{bash}
grep "number of SNPs:" /data/genetics_tmp/mapping_tests/results_NT/duplicate_filtering_AF_validation/VCF_dedup_30/validationIndelFilteredSNPs_Biallelic_500kb_ADP100.vcf_stats.txt
```

Number of SNPs in the Consensus data sets (per population inner join of Ind VCF and Pool VCF):
```{r}
paste0("Number of SNPs in consensus data set of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_dedup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Deduplication with fastp RD=30"
# print the results
cat(paste0(title,results_dedup[[1]], results_dedup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   hist2dmaf(dataset=x)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histMAFfreq(dataset=x, title=title)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x)
}, mc.cores = detectCores() - 1)
```

Number of SNPs in the Consensus data sets after MAF>0.05 (per population inner join of Ind VCF and Pool VCF):
```{r}
poplist <- poplist
for (i in 1:2){
  poplist[[i]][,c(4,5)][poplist[[i]][,c(4,5)] < 0.05] <- NA
  poplist[[i]] <- na.omit(poplist[[i]])
}

paste0("Number of SNPs in consensus data set (MAF>0.05) of population 305 = ", nrow(poplist[[1]]) )
paste0("Number of SNPs in consensus data set (MAF>0.05) of population 312 = ", nrow(poplist[[2]]) )
```

Statistics for AF
```{r}
# Apply the error_stats function to each element of poplist using parallel processing
results_markdup <- mclapply(poplist, function(x) {
  error_stats(x$alt_ind, x$alt_pool)
}, mc.cores = detectCores() - 1)

title <- "Deduplication with fastp RD=30 - MAF>0.05"
# print the results
cat(paste0(title,results_markdup[[1]], results_markdup[[2]]))
```

Plots
```{r fig.width=6, fig.height=6}
mclapply(poplist, function(x){
  hist2d(dataset=x, bins=36)
}, mc.cores = detectCores() - 1)

# mclapply(poplist, function(x){
#   histALTfreq(dataset=x, title=title, bins = 39)
# }, mc.cores = detectCores() - 1)

mclapply(poplist, function(x){
  rawdiffplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```

```{r fig.width=12, fig.height=6}
mclapply(poplist, function(x){
  AF_barplot(dataset=x, bins=37)
}, mc.cores = detectCores() - 1)
```