#Created by Lars Littmann for ACORN project, using Aquarel project data for testing, 26.06.2023

args <- commandArgs() #Intercept all the arguments that our bash script sends to R through the bash command Rscript.
#This places the arguments in a string we can late read from easily.

print(args)
setwd(args[8]) #Set the working directory to the output directory specified in the bash script. 

library(utils)

####Data entry####

#Start with the Pool data
PoolData<-read.table(args[6]) #Load the poolseq frequency data

PoolData$V7 <- gsub("[\\%]", "", PoolData$V7) #Remove the % signs that are in the column
PoolData$V7 <- as.numeric(gsub("[\\,]", ".", PoolData$V7)) #Make sure that the decimal indicator is a "." and the entries become numeric values.
names(PoolData) <- c('Chrom', 'Position', 'RefPool', 'AltPool', 'TotalReadsPool', 'AltReadsPool', 'AltFreqPool') #Rename the columns
PoolData$AltFreqPool <- (PoolData$AltFreqPool/100) #Change over from percentages to frequency


#Include Individual-based data
IndData<-read.table(args[7], header = FALSE, row.names = NULL) #Load the poolseq frequency data. Bit more unwieldy.
names(IndData) <- c('Chrom', 'Position', 'AlleleCountInd', 'PloidyInd', 'RefFreqInd', 'AltFreqInd') #Rename all columns
IndData$AltFreqInd <- (gsub("^.{0,2}", "", IndData$AltFreqInd)) #Make sure to remove the alt allele, leaving just the frequency in this column
IndData$RefFreqInd <- (gsub("^.{0,2}", "", IndData$RefFreqInd)) #Make sure to remove the ref allele, leaving just the frequency in this column
IndData$AltFreqInd<-as.numeric(IndData$AltFreqInd) #Force 
IndData$RefFreqInd<-as.numeric(IndData$RefFreqInd)

summary(IndData$AltFreqInd)

####Final Filtering####

#Pool data requires some final filtering, so that only entries with sufficient read depth and alt allele frequency are included.
PoolData<-PoolData[which(PoolData$TotalReadsPool>=175 & PoolData$AltFreqPool>0.01),] #THESE FILTER THRESHOLDS NEED REVISITING

####Merging####

#This line merges the up to now seperate dataframes by the chromosome/position pairs for which they both have valid data. ALL OTHER DATA IS DISREGARDED.
AllData<-merge(PoolData, IndData, by = c('Chrom','Position'))

####Linear Regression####

#Set up the linear model using the allele frequency found through individual sequencing as a predictor for the allele fequency found through pool sequencing.
AlleleFreqModel<-lm(AltFreqPool ~ AltFreqInd, AllData)

#Extract the R-squared value from the model and store it in a convenient string that can be used for figures.
RsquaredValue<-summary(AlleleFreqModel)$adj.r.squared #extract value
RsquaredString <-paste0('Adj. R-squared value = ', RsquaredValue) #Place in string

plot(AlleleFreqModel$residuals, pch = 16, col = "red", main = 'residuals of linear model') #Plot the residuals of the model. Mostly just a formality. No pattern in distribution.

#####Plotting#####

#Specify an outdirectory for the figure to go (remember, we are running this on a remote machine in a terminal)
PathOutDir<-args[8]

#Plot the regression. Very rudimentary plot for now, but can always be changed to a ggplot figure later (not my strength)
plot(AllData$AltFreqInd, AllData$AltFreqPool, 
     xlim = c(0,1.1), 
     ylim = c(0,1.1), 
     main = 'Regression between PoolSeq and IndSeq allele frequency', 
     xlab = 'Allele frequencies from IndSeq data', 
     ylab = 'Allele frequencies from PoolSeq data',
)
abline(AlleleFreqModel, col = 'red', lwd = 4) #Add the regression line, coefficient and intersect extracted from the model.
legend(0.4, 0.1, legend = paste0(RsquaredString)) #Print the R-squared value in the figure, by 'abusing' the legend function.

#NOTE: Plots generated by this script will automatically appear as a pdf in the working directory on the remote machine. 