---
title: "ACORN Pool-Seq PCAdapt outlier SNPs"
author: "Nikolaos Tourvas"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    code_download: yes
    code_folding: show
    theme: default
    toc: yes
    toc_float: yes
    keep_md: no
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

## Palettes
```{r}
cb_paired=c("#B1DF8A","#33A02C","#A6CEE3","#1F78B4","#FB9A99","#E31A1C",
            "#CAB2D6","#6A3D9A","#FDBF6F","#FF7F00","grey50","#B15928",
            "black")

wong=c("#000000","#E69F00","#56B4E9","#009E73","#CC79A7","#F0E442","#006699",
         "#D55E00")
```

## Libraries
```{r}
library(poolfstat)
library(pcadapt)
library(qvalue)
library(ggplot2)
library(dplyr)
library(stringr)
```

```{bash}
mkdir -p ../results/figs
```

## Load data
```{r}
pool209220 <- vcf2pooldata(vcf.file = "/mnt/results/VCF/Qrob_total_filter2.snp.vcf",
                           poolnames = c("209", "220"),
                           poolsizes = c(44, 47))
```

Verify chromosomes were loaded correctly
```{r}
unique(pool209220@snp.info$Chromosome)
```

Filter for coverage (not necessary)
```{r}
filtered <- vcf2pooldata(vcf.file = "/mnt/results/VCF/Qrob_total_filter2.snp.vcf",
     poolnames = c("209", "220"),
       poolsizes = c(44, 47),
     min.cov.per.pool = 45,
     max.cov.per.pool = 450,
     min.maf = 0.05)
```

Extract allele frequencies
```{r}
ref_freq <- t(filtered@refallele.readcount / filtered@readcoverage)
```

# Analysis
Keep only SNPs found in Chromosomes, NOT scaffolds!
```{r}
#get chromosome info
chrom <- filtered@snp.info
str(chrom)
dim(chrom)

# find rows of chromosome SNPS
chrom <- dplyr::filter(chrom, str_detect(Chromosome, "Qrob_Chr"))
# filter pcadapt input file
pcadapt_pre_input <- ref_freq[ ,1:nrow(chrom)]
```

```{r}
pcadapt_input <- read.pcadapt(pcadapt_pre_input, type = "pool")
```

Perform pcadapt analysis
PCAdapt offers a SNP thinning function, but it is not available for pooled
data as it is impossible to calculate LD
```{r}
res <- pcadapt(pcadapt_input, K=1, min.maf = 0.05)
#The same as res <- pcadapt(pcadapt_input,K=2)
summary(res)
```

To evaluate if LD might be an issue for our dataset, we will display the loadings 
(contributions of each SNP to the PC) and  evaluate if the loadings are clustered 
in a single or several genomic regions.
```{r fig.width=10, fig.height=10}
par(mfrow = c(2, 2))
for (i in 1:1)
  plot(res$loadings[, i], pch = 19, cex = .3, 
       col = rgb(red=0, green=0, blue=1, alpha = 0.2),
       ylab = paste0("Loadings PC", i))
```

Make a PCA plot
```{r}
plot(res, option = "scores", pop = c("209", "220"))
```

Investigate screeplot
Decide how many PCA axes to use
Hint: You can also decide based on PCA plots for axes 3,4 or more
```{r}
plot(res,option="screeplot")
```

Rerun pcadapt with the chosen number of PCA axes
```{r}
# res <- pcadapt(pcadapt_input, K=1, min.maf = 0.05)
# #The same as res <- pcadapt(pcadapt_input,K=2)
# summary(res)
```

list of outliers
```{r}
padj <- p.adjust(res$pvalues, method = "fdr")
alpha <- 0.01
outliers <- which(padj < alpha)
length(outliers)
```

```{r}
plot(res, option = "qqplot")
```

```{r}
hist(res$pvalues, xlab = "p-values", main = NULL, breaks = 50, col = "orange")
```

```{r}
plot(res, option = "stat.distribution")
```

Export outliers
```{r}
write.csv(x = outliers,
          file = "../results/VCF/pcadapt_outliers.csv",
          quote = FALSE, row.names = FALSE, col.names = FALSE)
```

Make Manhattan plot by chromosome
NA p values arise because of MAF filter
MAF filter works differently in poolfstat and pcadapt
In poolfstat MAF filter is global (as if the data consisted of one pop only)
PCAdapt MAF filter is applied per population (pool) basis
```{r fig.width=10, fig.height= 5}
pca_df <- as.data.frame(cbind(chrom, padj))
colnames(pca_df) <- c("chr", "position", "ref", "alt", "pvalues")
pca_df$chr <- gsub(pattern = "Qrob_Chr0", replacement = "", pca_df$chr)
pca_df$chr <- gsub(pattern = "Qrob_Chr", replacement = "", pca_df$chr)
pca_df$chr <- as.factor(reorder(pca_df$chr, sort(as.integer(pca_df$chr))))
str(pca_df)

manh_p <- ggplot(data = pca_df, aes(x = position, 
                                    y = -log10(pvalues),
                                    color = reorder(chr,
                                                    sort(as.numeric(chr))))) +
  facet_wrap(~chr, scales = "free_x", nrow=1, strip.position = "bottom") +
  geom_point(alpha = 1, size = 0.8) +
  geom_abline(intercept = -log10(sort(padj)[length(outliers)]), 
              slope = 0, linetype = "dashed") +
  scale_color_manual(values = rep(wong[-6], 3)) +
  theme_minimal() +
  theme(legend.position = "none",
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.spacing = unit(0, "lines")) +
  xlab("Chromosome") + ylab("-log10(p-value)")

manh_p + 
    theme(panel.grid.major.x = element_line(colour = NA,
      linetype = "blank"),
      panel.grid.minor = element_line(linetype = "blank"),
      plot.background = element_rect(fill = "white", color = "white"))

ggsave(filename = "Qrob_pcadapt-p-values-by-chr.pdf",
       path = "../results/figs", dpi=300,
       width = 10, height = 5)
```

```{r fig.width=10, fig.height=24}
manh_p2 <- ggplot(data = pca_df, aes(x = position / 10^6, 
                                    y = -log10(pvalues),
                                    color = chr)) +
  facet_wrap(~chr, scales = "free", ncol=1, strip.position = "top") +
  geom_point(alpha = 1, size = 0.8) +
  geom_abline(intercept = -log10(sort(padj)[length(outliers)]), 
              slope = 0, linetype = "dashed") +
  scale_color_manual(values = rep(wong[-6], 4)) +
  theme_bw() +
  theme(legend.position = "none",
        panel.grid.major.x = element_line(colour = NA, linetype = "blank"),
        panel.grid.minor = element_line(linetype = "blank"),
        panel.spacing = unit(0, "lines")) +
  xlab("Position (Mb)") + ylab("-log10(p-value)") +
  scale_x_continuous(labels = scales::comma)

manh_p2 + 
    theme(plot.background = element_rect(fill = "white", color = "white"))

ggsave(filename = "Qrob_pcadapt-p-values-by-chr2.pdf",
       path = "../results/figs", dpi=300,
       width = 10, height = 24)
```


